linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -4683,4 +4683,191 @@ private void flush() {
+      Function1<TSource, @Nullable TKey> sortKeySelector,
+      Comparator<TKey> sortComparator,
+      boolean all,
+      EqualityComparer<TSource> equalityComparer) {
+    return new AbstractEnumerable<TSource>() {
+      @Override public Enumerator<TSource> enumerator() {
+        return new MergeUnionEnumerator<>(
+            sources,
+            sortKeySelector,
+            sortComparator,
+            all,
+            equalityComparer);
+      }
+    };
+  }
+
+  /**
+   * Performs a union (or union all) of all its inputs (which must be already sorted),
+   * respecting the order.
+   * @param <TSource> record type
+   * @param <TKey> sort key
+   */
+  @SuppressWarnings("unchecked")
+  private static class MergeUnionEnumerator<TSource, @Nullable TKey>
+      implements Enumerator<TSource> {
+    private final List<Enumerator<TSource>> inputs;
+    private final TSource[] currentInputsValues;
+    private final Function1<TSource, @Nullable TKey> sortKeySelector;
+    private final Comparator<@Nullable TKey> sortComparator;
+    private TSource currentValue;
+    private int activeInputs;
+
+    // Set to control duplicates, only used if "all" is false
+    private final @Nullable Set<Wrapped<TSource>> processed;
+    private final @Nullable Function1<TSource, Wrapped<TSource>> wrapper;
+
+    private static final Object NOT_INIT = new Object();
+    private static final Object FINISHED = new Object();
+
+    @SuppressWarnings("method.invocation.invalid")
+    private MergeUnionEnumerator(
+        List<Enumerable<TSource>> sources,
+        Function1<TSource, @Nullable TKey> sortKeySelector,
+        Comparator<@Nullable TKey> sortComparator,
+        boolean all,
+        EqualityComparer<TSource> equalityComparer) {
+      this.sortKeySelector = sortKeySelector;
+      this.sortComparator = sortComparator;
+
+      if (all) {
+        this.processed = null;
+        this.wrapper = null;
+      } else {
+        this.processed = new HashSet<>();
+        this.wrapper = wrapperFor(equalityComparer);
+      }
+
+      final int size = sources.size();
+      this.inputs = new ArrayList<>(size);
+      for (Enumerable<TSource> source : sources) {
+        this.inputs.add(source.enumerator());
+      }
+
+      this.currentInputsValues = (TSource[]) new Object[size];
+      this.activeInputs = this.currentInputsValues.length;
+      this.currentValue = (TSource) NOT_INIT;
+
+      this.initEnumerators();
+    }
+
+    private void initEnumerators() {
+      for (int i = 0; i < this.inputs.size(); i++) {
+        this.moveEnumerator(i);
+      }
+    }
+
+    private void moveEnumerator(int i) {
+      final Enumerator<TSource> enumerator = this.inputs.get(i);
+      if (!enumerator.moveNext()) {
+        this.activeInputs--;
+        this.currentInputsValues[i] = (TSource) FINISHED;
+      } else {
+        this.currentInputsValues[i] = enumerator.current();
+      }
+    }
+
+    @SuppressWarnings("dereference.of.nullable")
```

Comment:
Does it mean there's NPE hidden in the method somewhere?

Reply:
No. CheckerFramework flags `wrapper.apply` as a possible NPE because `wrapper` can be theoretically null. However, `wrapper`  is null only when `processed` is null (and vice-versa), and since the method checks for `processed` nullability at the beginning, we can never get a NPE on  `wrapper.apply`

Reply:
If you intend to suppress a single case only, please use `SuppressWarnings` for a local variable declaration like in

```java
@SuppressWarnings("dereference.of.nullable")
final Wrapped<TSource> wrapped = this.wrapper.apply(value);
```

Reply:
Done